(Likely) false positive for unnecessary defer

```go
package main

import (
	"fmt"
	"sync"
)

type T struct {
	l    *sync.Mutex
	data []int
}

func (t *T) Index(i int) int {
	t.l.Lock()
	defer t.l.Unlock()
	return t.data[i]
}

func main() {
	t := T{l: &sync.Mutex{}, data: []int{1, 2, 3}}
	fmt.Println(t.Index(2))
}
```

Implementation of the index method is idiomatic and having this warning in this case isn't optimum.

```sh
 gocritic check -enable='unnecessaryDefer' main.go
./main.go:15:2: unnecessaryDefer: defer t.l.Unlock() is placed just before return
```

--- a/checkers/testdata/unnecessaryDefer/negative_tests.go
+++ b/checkers/testdata/unnecessaryDefer/negative_tests.go
@@ -53,3 +53,16 @@ func foo_8() int {
 	}
 	return 0
 }
+
+type sharedData struct {
+	value int
+}
+
+func (*sharedData) Lock()   {}
+func (*sharedData) Unlock() {}
+
+func issue941(d *sharedData) int {
+	d.Lock()
+	defer d.Unlock()
+	return d.value
+}

--- a/checkers/testdata/unnecessaryDefer/positive_tests.go
+++ b/checkers/testdata/unnecessaryDefer/positive_tests.go
@@ -56,3 +56,11 @@ func foo7() {
 	}
 	return
 }
+
+func returnConstExpr(s *sharedData) (string, bool, int) {
+	const foo = "12"
+	s.Lock()
+	/*! defer s.Unlock() is placed just before return */
+	defer s.Unlock()
+	return foo + "3", false, len(foo) + 1
+}

--- a/checkers/unnecessaryDefer_checker.go
+++ b/checkers/unnecessaryDefer_checker.go
@@ -4,10 +4,8 @@ import (
 	"go/ast"
 
 	"github.com/go-critic/go-critic/checkers/internal/astwalk"
-	"github.com/go-critic/go-critic/checkers/internal/lintutil"
 	"github.com/go-critic/go-critic/framework/linter"
 	"github.com/go-toolsmith/astfmt"
-	"github.com/go-toolsmith/astp"
 )
 
 func init() {
@@ -69,12 +67,11 @@ func (c *unnecessaryDeferChecker) checkDeferBeforeReturn(funcDecl *ast.BlockStmt
 			continue
 		}
 		explicitReturn = true
-		if lintutil.ContainsNode(retStmt, astp.IsCallExpr) {
+		if !c.isTrivialReturn(retStmt) {
 			continue
 		}
 		retIndex = i
 		break
-
 	}
 	if retIndex == 0 {
 		return
@@ -88,6 +85,19 @@ func (c *unnecessaryDeferChecker) checkDeferBeforeReturn(funcDecl *ast.BlockStmt
 			c.warn(deferStmt)
 		}
 	}
+}
+
+func (c *unnecessaryDeferChecker) isTrivialReturn(ret *ast.ReturnStmt) bool {
+	for _, e := range ret.Results {
+		if !c.isConstExpr(e) {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *unnecessaryDeferChecker) isConstExpr(e ast.Expr) bool {
+	return c.ctx.TypesInfo.Types[e].Value != nil
 }
 
 func (c *unnecessaryDeferChecker) warn(deferStmt *ast.DeferStmt) {

